### PR TITLE
v158: Attributes controller, correcting PHP notices when various Conf…

### PR DESCRIPTION
**Pull request by lat9**
…iguration :: Attribute Settings are set to 'false'.

Seeing logs similar to
```
22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_price_factor in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 252.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_price_factor_offset in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 253.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_price_factor_onetime in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 254.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_price_factor_onetime_offset in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 255.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_qty_prices in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 256.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_qty_prices_onetime in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 257.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_price_words in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 259.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_price_words_free in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 260.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_price_letters in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 261.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: attributes_price_letters_free in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 262.

[22-Apr-2022 18:02:30 Europe/Berlin] Request URI: /zc158/hoisT-PYL-teAch/index.php?cmd=attributes_controller&action=add_product_attributes&products_filter=25, IP address: 127.0.0.1
#1  require(C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php) called at [C:\xampp\htdocs\zc158\hoisT-PYL-teAch\index.php:11]
--> PHP Notice: Undefined index: img_dir in C:\xampp\htdocs\zc158\hoisT-PYL-teAch\attributes_controller.php on line 278.
```

... when one or more of these **_Configuration :: Attribute Settings_** are 'false:
image

Noting that these notices are also thrown by the v157 base.

### CODE REVIEW

File: admin/attributes_controller.php [hunk 1]
```diff
@@ -247,45 +247,69 @@
...
+              $attributes_price_words_free = (int)$_POST['attributes_price_words_free'];
+              $attributes_price_letters = (float)$_POST['attributes_price_letters'];
+              $attributes_price_letters_free = (int)$_POST['attributes_price_letters_free'];
+          } else {
+              $attributes_price_words = 0;
+              $attributes_price_words_free = 0;
+              $attributes_price_letters = 0;
+              $attributes_price_letters_free = 0;
+          }
+
```
Comment (proseLA) [inline admin/attributes_controller.php:283]: could not all of this code be simplified using the coalesce `??` operator?

**Review by proseLA — COMMENTED** [review]
as v158 requires a minimum of php 7; i think the coalesce operator would work better here.

File: admin/attributes_controller.php (hunk 1 above)
Comment (lat9) [reply admin/attributes_controller.php:283]: Probably, but that would have required changes to each line; it was more expedient to simply break out the condition.

Comment (lat9) [reply admin/attributes_controller.php:283]: ... as well as identifying the configuration settings that were introducing the 'conflict'.

Comment (proseLA) [reply admin/attributes_controller.php:283]: identifying the config settings are completely unnecessary.

Comment (lat9) [reply admin/attributes_controller.php:283]: If you want to rework each line to use the formatting you've specified, go for it!

**Comment by lat9** [comment]
FWIW, I've provided an update that addresses the source of the issue.  The rendering 'bits' needed to add hidden fields with the default values.

**Comment by scottcwilson** [comment]
@lat9 if you merge #4741 please close this one.

**Comment by proseLA** [comment]
FWIW, it's a bit of a thankless job, but i think this script is in serious need of refactoring.  it's currently over 2000 lines long, and there is so much code duplication for no good reason.  i look at this section of code right at the top:

https://github.com/zencart/zencart/blob/226cea58f31516c172eed2c298035796713a2afe/admin/attributes_controller.php#L64-L79

lines 71-73 and 77-79 seem exactly the same.  am i wrong?

line 64, how could `$products_filter` be set?

could we not use `$_REQUEST` for some of these things?

lines 242 - 263 and lines 382 to 403....  they look exactly the same.

i get it.  there are only a few of us working on this project...  potentially we find problems via working on clients sites, but perhaps we can do a little code cleanup as we go along....  instead of just fixing the errors that we find to address clients sites...

just my opinion....

**Pull request merged**